install_config/topics/sssd_for_ldap_failover_configure_apache.adoc

```diff
@@ -0,0 +1,194 @@
++
+----
+auth required pam_sss.so
+account required pam_sss.so
+----
++
+This configuration enables PAM, the pluggable authentication module, to use
+*pam_sss.so* to determine authentication and access control when an
+authentication request is issued for the *openshift* stack.
+
+. Edit the *_/etc/httpd/conf.modules.d/55-authnz_pam.conf_* file and uncomment
+ the following line:
++
+----
+LoadModule authnz_pam_module modules/mod_authnz_pam.so
+----
+
+. To configure the Apache *_httpd.conf_* file for remote basic authentication,
+create the *_openshift-remote-basic-auth.conf_* file in the 
+*_/etc/httpd/conf.d_* directory. Use the following template to provide your
+required settings and values:
++
+[IMPORTANT]
+====
+Carefully review the template and customize its contents to fit your 
+environment.
+====
++
+----
+LoadModule request_module modules/mod_request.so
+LoadModule php7_module modules/libphp7.so
+
+# Nothing needs to be served over HTTP.  This virtual host simply redirects to
+# HTTPS.
+<VirtualHost *:80>
+  DocumentRoot /var/www/html
+  RewriteEngine              On
+  RewriteRule     ^(.*)$     https://%{HTTP_HOST}$1 [R,L]
+</VirtualHost>
+
+<VirtualHost *:443>
+  # This needs to match the certificates you generated.  See the CN and X509v3
+  # Subject Alternative Name in the output of:
+  # openssl x509 -text -in /etc/pki/tls/certs/remote-basic.example.com.crt
+  ServerName remote-basic.example.com
+
+  DocumentRoot /var/www/html
+
+  # Secure all connections with TLS
+  SSLEngine on
+  SSLCertificateFile /etc/pki/tls/certs/remote-basic.example.com.crt
+  SSLCertificateKeyFile /etc/pki/tls/private/remote-basic.example.com.key
+  SSLCACertificateFile /etc/pki/CA/certs/ca.crt
+
+  # Require that TLS clients provide a valid certificate
+  SSLVerifyClient require
+  SSLVerifyDepth 10
+
+  # Other SSL options that may be useful
+  # SSLCertificateChainFile ...
+  # SSLCARevocationFile ...
+
+  # Send logs to a specific location to make them easier to find
+  ErrorLog logs/remote_basic_error_log
+  TransferLog logs/remote_basic_access_log
+  LogLevel warn
+
+  # PHP script that turns the Apache REMOTE_USER env var
+  # into a JSON formatted response that OpenShift understands
+  <Location /check_user.php>
+    # all requests not using SSL are denied
+    SSLRequireSSL
+    # denies access when SSLRequireSSL is applied
+    SSLOptions +StrictRequire
+    # Require both a valid basic auth user (so REMOTE_USER is always set)
+    # and that the CN of the TLS client matches that of the OpenShift master
+    <RequireAll>
+      Require valid-user
+      Require expr %{SSL_CLIENT_S_DN_CN} == 'system:openshift-master'
+    </RequireAll>
+    # Use basic auth since OpenShift will call this endpoint with a basic challenge
+    AuthType Basic
+    AuthName openshift
+    AuthBasicProvider PAM
+    AuthPAMService openshift
+    
+    # Store attributes in environment variables
+    LookupOutput Env
+    LookupUserAttr mail REMOTE_USER_MAIL
```

Comment:
@stuartchuan I have checked this manually by `SetEnv REMOTE_USER_MAIL test@test.com` which sets the email in the identity as expected.  Can you confirm that your mod lookup and SSSD set up is correct and that it is sending a non-empty `REMOTE_USER_MAIL` header?  You also need to make sure that `mail` is the attribute that stores the user's email.

I know it is at least partially working because I see this in your YAML:

```yaml
  extra:
    name: pm5 IPA
```

which likely means that either `mail` is the wrong attribute or SSSD is not configured to allow Apache to read it (see `Grant Apache permission to retrieve the email attribute`).

Please also provide the output of `oc get user -o yaml`.

Reply:
@enj all the setting following the docs, and how to check it is sending a non-empty REMOTE_USER_MAIL header?
The mail is the attribute that stored in the user:
--------------
1 user matched
--------------
  User login: pm5
  First name: pm5
  Last name: IPA
  Home directory: /home/pm5
  Login shell: /bin/sh
  Principal name: pm5@EXAMPLE.IO
  Principal alias: pm5@EXAMPLE.IO
  Email address: pm5@example.io
  UID: 1655400007
  GID: 1655400007
  Account disabled: False

# oc get user pm5 -o yaml
apiVersion: user.openshift.io/v1
fullName: pm5 IPA
groups: null
identities:
- sssd:pm5
kind: User
metadata:
  creationTimestamp: 2018-06-01T02:41:08Z
  name: pm5
  resourceVersion: "7602"
  selfLink: /apis/user.openshift.io/v1/users/pm5
  uid: 3d3037fa-6545-11e8-989f-fa163e724cb3

# oc get identity sssd:pm5 -o yaml
apiVersion: user.openshift.io/v1
extra:
  name: pm5 IPA
kind: Identity
metadata:
  creationTimestamp: 2018-06-01T02:41:08Z
  name: sssd:pm5
  resourceVersion: "7603"
  selfLink: /apis/user.openshift.io/v1/identities/sssd%3Apm5
  uid: 3d30b1f5-6545-11e8-989f-fa163e724cb3
providerName: sssd
providerUserName: pm5
user:
  name: pm5
  uid: 3d3037fa-6545-11e8-989f-fa163e724cb3

Reply:
It is unclear what attribute `Email address: pm5@example.io` uses.

I assume the debug logs will show envelope vars?

Reading https://www.adelton.com/apache/mod_lookup_identity may be helpful.

Reply:
The attribute is email in the IPA server.
I could find some error hint but still could not fix it.
[Fri Jun 01 04:53:03.600737 2018] [lookup_identity:debug] [pid 8378] mod_lookup_identity.c(445): [client **:47036] invoked for user pm
[Fri Jun 01 04:53:03.610369 2018] [lookup_identity:error] [pid 8378] [client **:47036] dbus call GetUserAttr returned value 0 instead of DBUS_TYPE_DICT_ENTRY

Reply:
@stuartchuan if `email` is the correct attribute in your setup, then you would have to update your configuration to use `email` instead of `mail` correct?  I find this surprising because IPA docs seem to indicate that the attribute is `mail`.  The error probably indicates that `LookupUserAttr mail REMOTE_USER_MAIL` is failing?

Reply:
@adelton can you help us here ?

Reply:
In IPA, the attribute for email address is `mail`.

The return value `0` seems like something is misconfigured on SSSD / SSSD ifp side.

You can verify the operation outside of Apache with

```
dbus-send --print-reply --system --dest=org.freedesktop.sssd.infopipe \
    /org/freedesktop/sssd/infopipe org.freedesktop.sssd.infopipe.GetUserAttr \
    string:pm5 array:string:mail
```

Reply:
# dbus-send --print-reply --system --dest=org.freedesktop.sssd.infopipe /org/freedesktop/sssd/infopipe org.freedesktop.sssd.infopipe.GetUserAttr string:pm5 array:string:mail
method return time=1528093813.798625 sender=:1.2787 -> destination=:1.2810 serial=23 reply_serial=2
   array [
   ]

When string setting to user email could get results,
# dbus-send --print-reply --system --dest=org.freedesktop.sssd.infopipe /org/freedesktop/sssd/infopipe org.freedesktop.sssd.infopipe.GetUserAttr string:pm5@example.io array:string:mail
method return time=1528091855.672691 sender=:1.2787 -> destination=:1.2795 serial=13 reply_serial=2
   array [
      dict entry(
         string "mail"
         variant             array [
               string "pm5@example.io"
            ]
      )
   ]

Reply:
When login by user email to the cluster, the mail field will present to the identity field.
```
# oc get user -o  yaml
apiVersion: v1
items:
- apiVersion: user.openshift.io/v1
  fullName: pm5 IPA
  groups: null
  identities:
  - sssd:pm5@example.io
  kind: User
  metadata:
    creationTimestamp: 2018-06-04T06:25:33Z
    name: pm5@example.io
    namespace: ""
    resourceVersion: "553293"
    selfLink: /apis/user.openshift.io/v1/users/pm5%40example.io
    uid: 15f75ffe-67c0-11e8-989f-fa163e724cb3
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""

# oc get identity -o yaml
apiVersion: v1
items:
- apiVersion: user.openshift.io/v1
  extra:
    email: pm5@example.io
    name: pm5 IPA
  kind: Identity
  metadata:
    creationTimestamp: 2018-06-04T06:25:33Z
    name: sssd:pm5@example.io
    namespace: ""
    resourceVersion: "553294"
    selfLink: /apis/user.openshift.io/v1/identities/sssd%3Apm5%40example.io
    uid: 15f7a78e-67c0-11e8-989f-fa163e724cb3
  providerName: sssd
  providerUserName: pm5@example.io
  user:
    name: pm5@example.io
    uid: 15f75ffe-67c0-11e8-989f-fa163e724cb3
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

Reply:
Not sure I understand but it looks like `pm5` is then not the identity SSSD would be able to resolve, and the fully qualified id has to be used.

What does `id pm5` and `id pm5@example.io` on the command line say?

Reply:
Should be the same user.
```
# id pm5
uid=1655400007(pm5) gid=1655400007(pm5@example.io) groups=1655400007(pm5@example.io)
# id pm5@example.io
uid=1655400007(pm5@example.io) gid=1655400007(pm5@example.io) groups=1655400007(pm5@example.io)
```

Reply:
In that case, it might be worth increasing the debug level in SSSD (or at least for its `ifp`) and check why over D-Bus, there is no reasonable reply for the short identifier (`pm5`).

Reply:
Here is the debug level 10 for ifp, is there any hint?
```
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sbus_dispatch] (0x4000): dbus conn: 0x7f4a0e9bfe40
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sbus_dispatch] (0x4000): Dispatching.
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sbus_signal_handler] (0x2000): Received D-Bus signal org.freedesktop.DBus.NameOwnerChanged
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sbus_signal_handler_got_caller_id] (0x0400): Got a signal from the bus..
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sbus_signal_name_owner_changed] (0x0400): Clearing UIDs cache
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sbus_dispatch] (0x4000): dbus conn: 0x7f4a0e9bfe40
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sbus_dispatch] (0x4000): Dispatching.
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sbus_message_handler] (0x2000): Received SBUS method org.freedesktop.sssd.infopipe.GetUserAttr on path /org/freedesktop/sssd/infopipe
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sbus_get_sender_id_send] (0x2000): Looking for identity of sender [:1.2849]
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sbus_add_timeout] (0x2000): 0x7f4a0e9ba5b0
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sbus_remove_timeout] (0x2000): 0x7f4a0e9ba5b0
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sbus_dispatch] (0x4000): dbus conn: 0x7f4a0e9bfe40
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sbus_dispatch] (0x4000): Dispatching.
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ifp_user_get_attr] (0x0040): GetUserAttr is deprecated, please consider switching to org.freedesktop.sssd.infopipe.Users.User interface
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ifp_user_get_attr] (0x0200): Looking up attributes of user [pm5] on behalf of 0
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sss_parse_name_for_domains] (0x0200): name 'pm5' matched without domain, user is pm5
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [cache_req_set_plugin] (0x2000): CR #5: Setting "User by name" plugin
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [cache_req_send] (0x0400): CR #5: New request 'User by name'
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [cache_req_process_input] (0x0400): CR #5: Parsing input name [pm5]
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sss_parse_name_for_domains] (0x0200): name 'pm5' matched without domain, user is pm5
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [cache_req_set_name] (0x0400): CR #5: Setting name [pm5]
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [cache_req_select_domains] (0x0400): CR #5: Performing a multi-domain search
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [cache_req_search_domains] (0x0400): CR #5: Search will check the cache and check the data provider
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [cache_req_validate_domain_type] (0x2000): Request type Any for domain example.io type POSIX is valid
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [cache_req_set_domain] (0x0400): CR #5: Using domain [example.io]
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [cache_req_prepare_domain_data] (0x0400): CR #5: Preparing input data for domain [example.io] rules
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [cache_req_search_send] (0x0400): CR #5: Looking up pm5@example.io
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [cache_req_search_ncache] (0x0400): CR #5: Checking negative cache for [pm5@example.io]
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sss_ncache_check_str] (0x2000): Checking negative cache for [NCE/USER/example.io/pm5@example.io]
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [cache_req_search_ncache] (0x0400): CR #5: [pm5@example.io] is not present in negative cache
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [cache_req_search_cache] (0x0400): CR #5: Looking up [pm5@example.io] in cache
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Added timed event "ltdb_callback": 0x7f4a0e9d1090

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Added timed event "ltdb_timeout": 0x7f4a0e9d2c90

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Running timer event 0x7f4a0e9d1090 "ltdb_callback"

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Destroying timer event 0x7f4a0e9d2c90 "ltdb_timeout"

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Ending timer event 0x7f4a0e9d1090 "ltdb_callback"

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Added timed event "ltdb_callback": 0x7f4a0e9cea20

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Added timed event "ltdb_timeout": 0x7f4a0e9d1090

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Running timer event 0x7f4a0e9cea20 "ltdb_callback"

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Destroying timer event 0x7f4a0e9d1090 "ltdb_timeout"

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Ending timer event 0x7f4a0e9cea20 "ltdb_callback"

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [cache_req_search_send] (0x0400): CR #5: Returning [pm5@example.io] from cache
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [cache_req_search_ncache_filter] (0x0400): CR #5: This request type does not support filtering result by negative cache
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [cache_req_create_and_add_result] (0x0400): CR #5: Found 1 entries in domain example.io
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [cache_req_done] (0x0400): CR #5: Finished: Success
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Added timed event "ltdb_callback": 0x7f4a0e9cea20

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Added timed event "ltdb_timeout": 0x7f4a0e9d3ec0

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Running timer event 0x7f4a0e9cea20 "ltdb_callback"

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Destroying timer event 0x7f4a0e9d3ec0 "ltdb_timeout"

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Ending timer event 0x7f4a0e9cea20 "ltdb_callback"

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Added timed event "ltdb_callback": 0x7f4a0e9cea20

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Added timed event "ltdb_timeout": 0x7f4a0e9c36d0

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Running timer event 0x7f4a0e9cea20 "ltdb_callback"

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Destroying timer event 0x7f4a0e9c36d0 "ltdb_timeout"

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Ending timer event 0x7f4a0e9cea20 "ltdb_callback"

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ifp_user_get_attr_handle_reply] (0x0080): Attribute mail not present or has no values
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sbus_dispatch] (0x4000): dbus conn: 0x7f4a0e9bfe40
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sbus_dispatch] (0x4000): Dispatching.
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sbus_signal_handler] (0x2000): Received D-Bus signal org.freedesktop.DBus.NameOwnerChanged
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sbus_signal_handler_got_caller_id] (0x0400): Got a signal from the bus..
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sbus_signal_name_owner_changed] (0x0400): Clearing UIDs cache
```

Reply:
And what's the debug output for the fully qualified username?

Reply:
The dubug output for user `pm5`

Reply:
I understand that. And we'd like to see the debug output when the `pm5@example.io` is used, so see where things start to differ.

Reply:
Also, please show the debug logs for the domain, for both cases.

Reply:
Here is the sssd.config
```
[domain/example.io]
cache_credentials = True
krb5_store_password_if_offline = True
ipa_domain = example.io
id_provider = ipa
auth_provider = ipa
access_provider = ipa
ldap_tls_cacert = /etc/ipa/ca.crt
ipa_hostname = www.example.io
chpass_provider = ipa
dyndns_update = True
ipa_server = www.example.io
dyndns_iface = ens3
dns_discovery_domain = example.io
realmd_tags = manages-system
default_shell = /bin/bash
use_fully_qualified_names = True
fallback_homedir = /home/%u@%d
ldap_user_extra_attrs = mail
debug_level = 6
[domain/default]
ldap_user_extra_attrs = mail
autofs_provider = ldap
ldap_schema = rfc2307bis
krb5_kpasswd = www.example.io:749
ldap_search_base = cn=users,cn=compat,dc=example,dc=io
krb5_server = www.example.io:88
id_provider = ldap
auth_provider = krb5
chpass_provider = krb5
krb5_store_password_if_offline = True
ldap_uri = ldap://www.example.io/
krb5_realm = EXAMPLE.IO
ldap_id_use_start_tls = True
cache_credentials = True
#ldap_tls_cacertdir = /etc/openldap/cacerts
ldap_tls_cacertdir = /etc/ipa
debug_level = 6
[sssd]
services = nss, sudo, pam, autofs, ssh, ifp, pac

domains = default, example.io
[nss]
homedir_substring = /home

[pam]

[sudo]

[autofs]

[ssh]

[pac]

[ifp]
user_attributes =  +mail
allowed_uids = apache, root
debug_level = 10
[secrets]

[session_recording]
```
The problem should be in the sssd configuration, there were two domains in the configuration, when login by `pm5` will direct to the default domain, and login by `pm5@example.io` will direct to example.io domain.
When I removed default domain from the configuration, only the `pm5@example.io` could login, and with mail in the identity. But when I removed example.io and left the defaut domain only, and change the default to example.io, both `pm5` and `pm5@example.io` could login, but no mail in the identity.

Would you point me how to modify the configuration file to let it working normally?

Reply:
Remove the `use_fully_qualified_names` from example domain and remove the default domain section completely?

Reply:
Definitely remove the double domains, you want only one.
About the use of fully qualified option or not, that is up to the deployment to decide but we can give guidance for people if they want to use short names.

Reply:
@enj suggestion was workable.
@adelton @simo5 @enj thank you all for the support.
No comments to the docs now, and all the steps in the guide were verified.